Framework/Algorithms/src/FindPeaksConvolve.cpp

```diff
@@ -0,0 +1,459 @@
+    }
+  }
+  return err_str;
+}
+
+void FindPeaksConvolve::storeClassProperties() {
+  m_inputDataWS = getProperty("InputWorkspace");
+  m_createIntermediateWorkspaces = getProperty("CreateIntermediateWorkspaces");
+  m_findHighestDatapointInPeak = getProperty("FindHighestDataPointInPeak");
+  m_iOverSigmaThreshold = getProperty("IOverSigmaThreshold");
+  m_mergeNearbyPeaks = getProperty("MergeNearbyPeaks");
+  // can we assume all spectra are either centred or not?
+  if (m_inputDataWS->x(0).size() != m_inputDataWS->y(0).size()) {
+    m_centreBins = true;
+  } else {
+    m_centreBins = false;
+  }
+
+  int startWSIndex = getProperty("StartWorkspaceIndex");
+  int endWSIndex = getProperty("EndWorkspaceIndex");
+  int specCount = static_cast<int>(m_inputDataWS->getNumberHistograms());
+  if (startWSIndex == EMPTY_INT()) {
+    startWSIndex = 0;
+  }
+  if (endWSIndex == EMPTY_INT()) {
+    endWSIndex = specCount - 1;
+  }
+  m_specCount = endWSIndex - startWSIndex + 1;
+  m_specNums.resize(m_specCount);
+  std::iota(std::begin(m_specNums), std::end(m_specNums), startWSIndex);
+  m_peakResults.resize(m_specCount);
+}
+
+void FindPeaksConvolve::performConvolution(const size_t dataIndex) {
+  const auto specNum{m_specNums[dataIndex]};
+  const HistogramData::HistogramX *xData{&m_inputDataWS->x(specNum)};
+  std::pair<const size_t, const bool> kernelBinCount{getKernelBinCount(xData)};
+  if (kernelBinCount.second) {
+    g_log.error("The kernel size for spectrum " + std::to_string(m_specNums[dataIndex]) +
+                " exceeds the range of the x axis. Please reduce the peak extent.");
+  } else {
+    const Tensor1D kernel{createKernel(static_cast<int>(kernelBinCount.first))};
+
+    const auto binCount{m_inputDataWS->getNumberBins(specNum)};
+    // Edge handling is performed by padding the input data with 0 values. Each convolution requires a padding of kernel
+    // size + 1. The 1st conv is performed with a kernel of size n, the second size n/2. The resultant pad is split
+    // either side of the data.
+    const double paddingSize = (std::ceil(static_cast<double>(kernelBinCount.first) * 1.5) - 2) / 2;
+    const TensorMap_const yData{&m_inputDataWS->y(specNum).front(), binCount};
+    Eigen::array<std::pair<double, double>, 1> paddings{
+        std::make_pair(std::ceil(paddingSize), std::floor(paddingSize))};
+    const auto yData_padded{yData.pad(paddings)};
+    const Eigen::array<ptrdiff_t, 1> dims({0});
+    const Tensor1D yConvOutput{yData_padded.convolve(kernel, dims)};
+
+    const auto eData{TensorMap_const{&m_inputDataWS->e(specNum).front(), binCount}.pad(paddings)};
+    const Tensor1D eConvOutput{eData.square().convolve(kernel.square(), dims).sqrt()};
+
+    const Tensor1D smoothKernel{
+        createSmoothKernel(static_cast<size_t>(std::ceil(static_cast<double>(kernelBinCount.first) / 2.0)))};
+    const Tensor1D iOverSigConvOutput{(yConvOutput / eConvOutput).convolve(smoothKernel, dims)};
```

Comment:
It is possible (particularly for low stats data) that the error convolution may include some zeros. In this case the corresponding y value should also be 0 - so it depends on how C++ handles zero division (in a numpy array this would produce NaN). How does Eigen handle this, and should we replace non-finite values with 0?

Reply:
Good point - I will check how Eigen behaves and get back.

Reply:
dividing by 0 produces `inf`, so I've added a step to change this to 0.